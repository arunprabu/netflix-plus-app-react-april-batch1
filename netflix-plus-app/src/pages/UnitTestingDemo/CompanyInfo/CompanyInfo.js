import PropTypes from 'prop-types';
import { useState } from 'react';

const CompanyInfo = ({ foundedYear }) => {
  const [countryName, setCountryName] = useState('USA');

  const handleCountryNameChange = (event) => {
    console.log(event.target.value);
    setCountryName(event.target.value);
  };

  return (
    <div>
      <h2>Company Info | Testing JSX, Props, CSS, Events, States and Snapshot</h2>
      <h3 data-testid="companyName">Company Name: Cognizant</h3>
      <p data-testid="foundedYear" style={{ color: '#ff0000', fontSize: '20px' }}>
        Founded Year: {foundedYear}
      </p>

      <input
        type="text"
        value={countryName}
        placeholder="Enter Your Country Name"
        onChange={handleCountryNameChange}
      />
      <p data-testid="visitWebsite">
        Please visit Cognizant {countryName} Website
      </p>
    </div>
  );
};

CompanyInfo.propTypes = {
  foundedYear: PropTypes.string
};

export default CompanyInfo;
