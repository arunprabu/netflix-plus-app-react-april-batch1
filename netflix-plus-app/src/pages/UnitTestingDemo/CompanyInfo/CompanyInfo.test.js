import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CompanyInfo from './CompanyInfo';

// TEST SUITE = group of related tests or test specs
describe('CompanyInfo', () => {
  // test spec
  it('has Cognizant as company name', () => {
    render(<CompanyInfo />);
    const companyName = screen.getByTestId('companyName');
    // console.log(companyName.textContent);
    expect(companyName.textContent).toBe('Company Name: Cognizant');
  });

  // testing props
  it('receives foundedYear prop and displays in JSX', () => {
    // when you are testing props
    // -- always render the comp with necessary props
    render(<CompanyInfo foundedYear="2000" />);
    const foundedYear = screen.getByTestId('foundedYear');
    expect(foundedYear.textContent).toBe('Founded Year: 2000');
  });

  // TODO: receives headquarters prop and displays in JSX
  // in browser I want to see this 'Headquarters: Teaneck, New Jersey, U.S'

  // Testing Inline Styles
  it('has company founded year in red color text and fontSize 20px', () => {
    render(<CompanyInfo foundedYear="2000" />);
    const foundedYear = screen.getByTestId('foundedYear');
    // [RECOMMENDED]: to use colors in either rgb format or hexa in both comp and test
    expect(foundedYear).toHaveStyle({
      color: 'rgb(255, 0, 0)',
      fontSize: '20px'
    });
  });

  // TODO: Testing External Css Class
  /*
   * testing whether a button has css class btn-primary
   * the button should have the text - JOIN COGNIZANT
   * use data-testid
   * you can have multiple expectations
   */

  // Let's find out whether the input has the right placeholder
  it('has an input with `Enter Your Country Name` as placeholder text', () => {
    render(<CompanyInfo />);
    // finding input element
    const countryInput = screen.getByPlaceholderText('Enter Your Country Name');
    expect(countryInput).toBeTruthy();
  });

  it('displays countryName USA by default and updates JSX onChange of country input', () => {
    render(<CompanyInfo />);
    // finding input element
    const countryInput = screen.getByPlaceholderText('Enter Your Country Name');
    console.log(countryInput.value);
    expect(countryInput.value).toBe('USA');

    const mockEventObject = {
      target: {
        value: 'Japan'
      }
    };

    // now checking whether the onChange event handler is working or not
    // as the comp is rendered in test runner -- we can't use mouse or keyboard as input devices
    // so, let's trigger the event thru program
    fireEvent.change(countryInput, mockEventObject);
    // finally let's check if the input element is showing the change
    expect(countryInput.value).toBe('Japan');

    expect(screen.getByTestId('visitWebsite')).toHaveTextContent(
      'Please visit Cognizant Japan Website'
    );
  });

  // Snapshot Testing
  it('has right CompanyInfo snapshot with all requirements completed', () => {
    // to take snapshot you need to use react-test-renderer // npm i react-test-renderer -D
    // Taking snapshot and converting it into JSON
    // [RECOMMENDED]: Take snapshot with all possible props as well as props children
    const snapshotInJson = renderer.create(<CompanyInfo foundedYear="2006"/>).toJSON();
    // Let's assert
    expect(snapshotInJson).toMatchSnapshot();
  });
});
