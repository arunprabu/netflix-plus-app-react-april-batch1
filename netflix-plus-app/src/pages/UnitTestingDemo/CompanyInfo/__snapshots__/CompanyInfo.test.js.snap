// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyInfo has right CompanyInfo snapshot with all requirements completed 1`] = `
<div>
  <h2>
    Company Info | Testing JSX, Props, CSS, Events, States and Snapshot
  </h2>
  <h3
    data-testid="companyName"
  >
    Company Name: Cognizant
  </h3>
  <p
    data-testid="foundedYear"
    style={
      Object {
        "color": "#ff0000",
        "fontSize": "20px",
      }
    }
  >
    Founded Year: 
    2006
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter Your Country Name"
    type="text"
    value="USA"
  />
  <p
    data-testid="visitWebsite"
  >
    Please visit Cognizant 
    USA
     Website
  </p>
</div>
`;
